### PR TITLE
Switch off camera when user is not focused on WorkAdventure windows

### PR DESCRIPTION
**Pull request by gparant**
Feature to switch off camera when user is not focused on WorkAdventure windows after 10 seconds.

### CODE REVIEW

**Review by moufmouf — CHANGES_REQUESTED** [review]
The idea is good, but we should probably improve the code a bit.

I'd say we must not disable the camera / microphone if the user is already speaking.
Just imagine: the user is speaking with someone, he enables screen sharing (because he wants to share his screen....) and boom, after 10 seconds of switching to the application he is sharing, the video signal is lost.

Also, if the user is using 2 screens (one with WorkAdventure on a secondary screen and one with his work), maybe he does want WorkAdventure to stay up (the focus is lost but WorkAdventure is visible). So maybe we want to check whether the window is *visible* rather than focused.

The good news: I know how to do it :D
Phaser animations are  based on "RequestAnimationFrame" (RAF). Each time a frame is requested, the "update()" method of the scene is called. When the tab is not visible at all, the browser stops calling "RequestAnimationFrame". So the "update()" method of the scene is not called any more.

By putting a "setInterval" in the scene, called once every 1-2 seconds and comparing the current time with the last time the update() method was called, we could be notified of the screen being hidden. And shut down the webcam at this very moment.

What do you think?

File: front/src/WebRtc/MediaManager.ts [hunk 1]
```diff
@@ -565,7 +631,33 @@ export class MediaManager {
         mainContainer.appendChild(divReport);
     }
 
+    //ping always 30 seconds data on stream
```
Comment (moufmouf) [inline front/src/WebRtc/MediaManager.ts:634]: Add this comment:

```
/**
 * For some reasons, the microphone muted icon or the stream is not always up to date.
 * Here, every 30 seconds, we are "reseting" the streams and sending again the constraints to the other peers via the data channel again (see SimplePeer::pushVideoToRemoteUser)
 */
```

**Comment by jacquesbh** [comment]
Does it cut the microphone as well?

Also, will the video come back when the tab is active again?

@moufmouf  The idea in #261 was to disable camera and microphone if no circle and no jitsi :).

We are full time on WOKA with some colleagues, like all day long. We don't share screen, we are just talking to each other. And having the video is a bonus when we go to the WOKA tab.

I would love to discuss this!

**Comment by moufmouf** [comment]
Hey @gparant ,

Just tested this. There are a lot of weird issues.
First, users seem to be disconnected / reconnected every 30 seconds from the server.

And when one user leaves, the server crashes with this stacktrace:

```
/usr/src/app/dist/src/Services/IoSocketHelpers.js:52
        ws.close();
           ^
Invalid access of closed uWS.WebSocket/SSLWebSocket.
(Use `node --trace-uncaught ...` to show where the exception was thrown)
```

Also, we need to make sure that if we shut down the mic and the cam, we can start them again without having to ask for special authorizations (especially in Firefox that is a bit picky about those authorizations).

@jacquesbh comment is also interesting (if it is technically possible). Today, a user told me she cannot use WorkAdventure and Microsoft Teams at the same time. Not really surprising since both apps require the webcam, but having WorkAdventure release its hold on the webcam could be great for compatibility with other apps.
We still need to make sure we can restart the webcam easily when it has been switched off, which seems as far as I can tell easier in Chrome than in Firefox.

**Comment by moufmouf** [comment]
@gparant Just did some more tests.

The first issue I'm talking about (users seem to be disconnected / reconnected every 30 seconds from the server) seems to happen on the "develop" branch too, so this is not related to this PR, sorry.

**Comment by moufmouf** [comment]
@gparant Sorry, the PR is not working as expected yet.

1- I connect to the game
2- I switch tab
3- I wait ~20 seconds
=> The light of the webcam is still turned on
4- When I come back to the game tab, I can see that the video was indeed removed and that it is enabled again (after a few milliseconds). But the light of the webcam is still on, so I guess during this whole time, the "getUserMedia" call is not properly released.

**Comment by gparant** [comment]
@dan I don't know why on your environment, the light still turned, but you could check on the log that you getUserMedia is released.

I'm not sure that we could to have a real impact on the "webcam light environment". For exemple with my Mac, when I siwitch off webcam on chrome, the light still turned on but when I use Firefox and I turn off webcam, the light is turned off. It's wird !

I think that my development is correct, I change condition and call getUserMedia to have new property each time.

**Pull request merged**